  <div class="alert alert-info">
  Привет, Кирилл! Меня зовут Светлана и я буду проверять твой проект. Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе. Предлагаю общаться на «ты». Но если это не удобно - дай знать, и мы перейдем на «вы».

  <div class="alert alert-success">
  <b>👍 Успех:</b> Зелёным цветом отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах.
  </div>
  <div class="alert alert-warning">
  <b>🤔 Рекомендация:</b> Жёлтым цветом выделено то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас (однако это не обязательно).
  </div>
  <div class="alert alert-danger">
  <b>😔 Необходимо исправить:</b> Красным цветом выделены комментарии, без исправления которых, я не смогу принять проект :(
  </div>
  <div class="alert alert-info">
  <b>👂 Совет:</b> Какие-то дополнительные материалы
  </div>
  Давай работать над проектом в диалоге: если ты что-то меняешь в проекте по моим рекомендациям — пиши об этом.
  Мне будет легче отследить изменения, если ты выделишь свои комментарии:
  <div class="alert alert-info"> <b>🎓 Комментарий студента:</b> Например, вот так.</div>
  Пожалуйста, не перемещай, не изменяй и не удаляй мои комментарии. Всё это поможет выполнить повторную проверку твоего проекта быстрее.
   </div>

# Проект по SQL

<div style="border:solid Purple 2px; padding: 40px" > 
    
<u><b> Описание проекта:</b></u>
    
Коронавирус застал мир врасплох, изменив привычный порядок вещей. На какое-то время жители городов перестали выходить на улицу, посещать кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.
Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.
    
<u><b> Задачи:</b></u>    

<li>Посчитать, сколько книг вышло после 1 января 2000 года;</li>
<li>Для каждой книги посчитать количество обзоров и среднюю оценку;</li>
<li>Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;</li>
<li>Определить автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;</li>
<li>Посчитайть среднее количество обзоров от пользователей, которые поставили больше 48 оценок.</li>
    
<u><b>Описание данных</b></u> 

Таблица books 

Содержит данные о книгах:

<li>book_id — идентификатор книги;</li>
<li>author_id — идентификатор автора;</li>
<li>title — название книги;</li>
<li>num_pages — количество страниц;</li>
<li>publication_date — дата публикации книги;</li>
<li>publisher_id — идентификатор издателя.</li>
    
    
Таблица authors

Содержит данные об авторах:

<li>author_id — идентификатор автора;</li>
<li>author — имя автора.</li>
    
Таблица publishers

Содержит данные об издательствах:

<li>publisher_id — идентификатор издательства;</li>
<li>publisher — название издательства;</li>
    
Таблица ratings

Содержит данные о пользовательских оценках книг:

<li>rating_id — идентификатор оценки;</li>
<li>book_id — идентификатор книги;</li>
<li>username — имя пользователя, оставившего оценку;</li>
<li>rating — оценка книги.</li>
Таблица reviews

Содержит данные о пользовательских обзорах на книги:

<li>review_id — идентификатор обзора;</li>
<li>book_id — идентификатор книги;</li>
<li>username — имя пользователя, написавшего обзор;</li>
<li>text — текст обзора.</li>

In [1]:
# импортируем библиотеки
import pandas as pd
from IPython.display import display
from sqlalchemy import text, create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [3]:
def q_exec(query='SELECT version()', show=False):
    '''
    Функция принимает запрос на языке sql и возвращает объект типа pandas DataFrame
    ## Parameters:
    `query` - запрос в виде строки. по умолчанию "SELECT version()"
    `show` - вывести сразу результат на экран без присваивания результата работы функции переменной, по умолчанию `False`
    '''
    if show:
        return display(pd.io.sql.read_sql(query, con = engine))
    else:
        return pd.io.sql.read_sql(query, con = engine)

In [4]:
# напишем функцию для сбора информации о датафрейме

def df_info(df=pd.DataFrame()):
    '''
    Функция выводит на экран 5 случайных строк и 
    общую информацию о датафрейме
    ## Parameters:
    `df` - объект pandas.DataFrame
    '''
    display(df.sample(5))
    print('-----------')
    print(df.info(), 'Явных дубликатов - ', df.duplicated().sum())

In [5]:
# таблица books
books_query = 'SELECT * FROM books'

df_info(q_exec(books_query))

book_id  author_id                                              title  \
356      357        314  Into Thin Air: A Personal Account of the Mount...   
508      509        117  Peter and the Starcatchers (Peter and the Star...   
723      724        505        The Fires of Heaven (The Wheel of Time  #5)   
449      450        590  Mountains Beyond Mountains: The Quest of Dr. P...   
369      370        631                                      Julius Caesar   

     num_pages publication_date  publisher_id  
356        368       1999-10-19            16  
508        452       2006-05-01            92  
723        912       1994-07-04           198  
449        333       2004-08-31           242  
369        128       2002-10-17           204

-----------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None Явных дубликатов -  0


 <div class="alert alert-danger">
 <s><b>😔 Необходимо исправить:</b> Функция вывода запроса куда то пропала и тетрадка не запускается</s>
 </div>

<div class="alert alert-info"> <b>🎓 Комментарий студента:</b> Замечание устранено.</div>

In [6]:
# таблица authors
authors_query = 'SELECT * FROM authors'

df_info(q_exec(authors_query))

author_id              author
618        619      William Gibson
527        528        Sarah Dessen
245        246  Jacquelyn Mitchard
280        281       Jen Lancaster
136        137         Donna Tartt

-----------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
None Явных дубликатов -  0


In [7]:
# таблица publishers
publishers_query = 'SELECT * FROM publishers'

df_info(q_exec(publishers_query))

publisher_id                   publisher
263           264                   Shambhala
254           255             Scholastic Inc.
335           336  Workman Publishing Company
21             22             Arrow Books Ltd
89             90               Digireads.com

-----------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
None Явных дубликатов -  0


In [8]:
# таблица ratings
ratings_query = 'SELECT * FROM ratings'

df_info(q_exec(ratings_query))

rating_id  book_id        username  rating
6353       6354      976         dharris       5
5045       5046      779        lorichen       5
3219       3220      505  carolrodriguez       3
5409       5410      837    fieldsangela       4
6186       6187      948     brentpowers       5

-----------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
None Явных дубликатов -  0


In [9]:
# таблица reviews
reviews_query = 'SELECT * FROM reviews'

df_info(q_exec(reviews_query))

review_id  book_id     username  \
2338       2337      839      fharris   
1822       1821      659  kristykirby   
1122       1123      413       ryan01   
2197       2197      792    richard89   
1138       1138      417  robertwells   

                                                   text  
2338  Method much plant simply. Go challenge word tend.  
1822  Guy company spend light fly recognize. Phone m...  
1122  Black indicate new meeting. Down relationship ...  
2197  Apply own radio phone do couple speech. Pm wha...  
1138  Admit leave time just rise. Son nothing gun. T...

-----------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
None Явных дубликатов -  0


<div style="background-color:aliceblue;padding:25px;"> Промежуточный вывод:
в нашем распоряжении 5 таблиц:
<li>books - 1000 книг, типы данных соответствуют определнию (кроме publication_date)</li>
<li>authors - 636 авторов книг</li>
<li>publishers - 340 издателей</li>
<li>ratings - 6456 оценок пользователей</li>
<li>reviews - 2793 пользовательских обзора</li>
<li>пропусков и явных дубликатов нет.</li>

 <div class="alert alert-success">
 <b>👍 Успех:</b> Есть описание задачи, импортированы все нужные библиотеки, создано подключение к БД, просмотрены таблицы
 </div>

## Подсчет количества книг, вышедших после 1 января 2000 года

In [10]:
q1 = '''
        SELECT count(book_id) 
        FROM books WHERE publication_date > '2000-01-01';
    '''

answer_1 = q_exec(q1)

answer_1 = int(answer_1.values[0])

answer_1

819

<div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

## Подсчет количества обзоров и средней оценки для каждой книги

In [11]:
q2 = '''
        WITH
                reviews_qty
        AS

        (SELECT DISTINCT
                b.title,
                b.book_id,
                COUNT(rw.review_id) AS review_qty
                
        FROM books b
            LEFT JOIN reviews rw
            USING(book_id)
  
        GROUP BY b.title, b.book_id
        ORDER BY 2 desc)

        SELECT  
                rq.title,
                rq.review_qty,
                AVG(r.rating)
        FROM reviews_qty rq
        LEFT JOIN ratings r
        USING(book_id)
        GROUP BY rq.title, rq.review_qty
        ORDER BY 2 desc
'''

answer_2 = q_exec(q2)

answer_2.head()

title  review_qty       avg
0                            Twilight (Twilight  #1)           7  3.662500
1                The Hobbit  or There and Back Again           6  4.125000
2  Harry Potter and the Chamber of Secrets (Harry...           6  4.287500
3                                    Eat  Pray  Love           6  3.395833
4  The Lightning Thief (Percy Jackson and the Oly...           6  4.080645

Больше всего обзоров получила книга Twilight (Twilight #1) — 7 штук, а вот рейтинг у нее — 3.66/10.

 <div class="alert alert-warning">
 <b>🤔 Рекомендация:</b> Выводить нужно полный результат запроса, а не частично
 </div>

## Определить издательство, которое выпустило наибольшее число книг толще 50 страниц

In [12]:
q3 = '''
        SELECT
                p.publisher,
                COUNT(p.publisher_id) AS publisher_qty
                
        FROM books b
            LEFT JOIN publishers p
            USING(publisher_id)
            
        WHERE
                b.num_pages > 50
                
        GROUP BY p.publisher_id
        ORDER BY 2 DESC
        LIMIT 1;      
'''

answer_3 = q_exec(q3)

answer_3 = answer_3['publisher'].values[0]

answer_3

'Penguin Books'

 <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

## Определить автора с самой высокой средней оценкой книг - учитываются только книги с 50 и более оценками

In [16]:
q4 = '''
        SELECT
                a.author,
                AVG(rating) AS average_rating
                
        FROM books b
            LEFT JOIN authors a
            USING(author_id)
            LEFT JOIN ratings r
            USING(book_id)   
             
        GROUP BY a.author
        HAVING COUNT(book_id) > 50
        ORDER BY 2 DESC
        LIMIT 100
'''

answer_4 = q_exec(q4)

answer_4 = answer_4['author'].values[0]

answer_4


'J.K. Rowling/Mary GrandPré'

In [15]:
q4 = '''WITH book_ratings AS (SELECT r.book_id,
                                        AVG(r.rating) AS avg_rating,
                                        COUNT(r.rating) AS total_ratings
                                 FROM ratings r
                                 GROUP BY r.book_id
                                 HAVING COUNT(r.rating) >= 50),
           author_average_ratings AS (SELECT a.author_id,
                                             SUM(br.total_ratings) AS total_ratings,
                                             AVG(br.avg_rating) AS avg_rating
                                      FROM authors a
                                      JOIN books b ON a.author_id = b.author_id
                                      JOIN book_ratings br ON b.book_id = br.book_id
                                      GROUP BY a.author_id)
           SELECT a.author,
                  aar.total_ratings,
                  aar.avg_rating
           FROM authors a
           JOIN author_average_ratings aar ON a.author_id = aar.author_id
           ORDER BY aar.avg_rating DESC
           LIMIT 1'''
answer_4 = q_exec(q4)

answer_4 = answer_4['author'].values[0]

answer_4

'J.K. Rowling/Mary GrandPré'

 <div class="alert alert-danger">
<s> <b>😔 Необходимо исправить:</b> Получилось не совсем верно, ты не учитываешь книги, где ровно 50 оценок, и фильтр HAVING призван отфильтровать книги с оценками 50 и более, а данные ты группируешь по автору. В итоге фильтр фильтрует не книги, а авторов, и в такой группировке почти у каждого автора будет 50 и более оценок, поэтому рейтинг будет рассчитан неверно - в него попадут лишние книги, которые не проходят по условию с количеством оценок</s>
 </div>

<div class="alert alert-info"> <b>🎓 Комментарий студента:</b> Замечание устранено. Автор с самой высокой средней оценкой книг (с 50 и более оценками): J.K. Rowling</div>

 <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

## Подсчет среднего количества обзоров от пользователей, которые поставили больше 50 оценок

In [14]:
q5 = '''
        WITH 
        
        users_above_48_rates AS
        
        (SELECT
                r.username,
                COUNT(r.username)
        FROM
                ratings r
        LEFT JOIN books b
        USING(book_ID)
        GROUP BY r.username
        HAVING COUNT(r.username) > 48)

        SELECT 
                COUNT(username) / COUNT(DISTINCT username)
        
        FROM users_above_48_rates ur
                
        LEFT JOIN reviews rw
        USING(username)
             
'''

answer_5 = q_exec(q5)

answer_5 = int(answer_5.values[0])

answer_5

24

 <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

## Выводы

Мы загрузили данные каждой из таблиц, сделали SQL-запросы и получили ответы на 5 вопросов.

Обзор данных:

в нашем распоряжении 5 таблиц:
<li>books - 1000 книг, типы данных соответствуют определнию (кроме publication_date)</li>
<li>authors - 636 авторов книг,</li>
<li>publishers - 340 издателей,</li>
<li>ratings - 6456 оценок пользователей,</li>
<li>reviews - 2793 пользовательских обзора.</li>
<li>пропусков и явных дубликатов нет.</li>

Результаты каждой задачи:
<li>819 книг вышло после 1 января 2000 года,</li>
<li>количество обзоров и средней оценки для каждой книги хранится в фрейме answer_2, больше всего обзоров получила книга Twilight (Twilight #1) — 7 штук, а вот рейтинг у нее — 3.7/10.</li>
<li>Penguin Books - издательство, которое выпустило наибольшее число книг толще 50 страниц,</li>
<li>J.K. Rowling/Mary GrandPré - автор с самой высокой средней оценкой книг (учитывались только книги с 50 и более оценками),</li>
<li>24 - среднее количество обзоров от пользователей, которые поставили больше 48 оценок.</li>

<div class="alert alert-success">
<b>👍 Успех:</b> Хорошая работа, поздравляю с окончанием курса! Рекомендую еще пройти бесплатный курс от практикума по <a href='https://practicum.yandex.ru/ycloud-datalens/'> DataLens</a>
   Дополнительную информацию по SQL  можно посмотреть и потренироваться <a href='https://sql-ex.ru'> здесь</a>, <a href='https://pgexercises.com'> здесь</a>, <a href='https://sql-academy.org'> здесь</a>, <a href='https://stepik.org/course/63054/info'> здесь</a>, <a href='https://Itresume.ru'> здесь</a>
</div>